File: lib/fulink/tyronProEngine.ts
/**
 * 타이롱 PRO 엔진 포팅 (assets/lib logic/tyron_pro_engine.dart)
 * 꼬리 흡수 + RSI 모멘텀 + 볼륨 스파이크 → bias, confidence, path
 */

import type { Candle } from '@/types';

export type TyronProResult = {
  bias: 'LONG' | 'SHORT' | 'NEUTRAL';
  confidence: number;
  absorbBull: boolean;
  absorbBear: boolean;
  reasons: string[];
  pathMain: number[];
  pathAlt: number[];
};

function atr(candles: Candle[], len: number): number {
  if (candles.length < len + 2) return 0;
  const start = candles.length - len;
  let sum = 0;
  for (let i = start; i < candles.length; i++) {
    const cur = candles[i]!;
    const prevClose = candles[i - 1]!.close;
    const tr = Math.max(cur.high - cur.low, Math.abs(cur.high - prevClose), Math.abs(cur.low - prevClose));
    sum += tr;
  }
  return sum / len;
}

function rsi(candles: Candle[], len: number): number {
  if (candles.length < len + 2) return 50;
  const start = candles.length - len;
  let gain = 0, loss = 0;
  for (let i = start; i < candles.length; i++) {
    const diff = candles[i]!.close - candles[i - 1]!.close;
    if (diff >= 0) gain += diff;
    else loss += -diff;
  }
  if (gain === 0 && loss === 0) return 50;
  if (loss === 0) return 100;
  const rs = gain / loss;
  return 100 - 100 / (1 + rs);
}

function buildPath(pathLen: number, atrVal: number, price: number, score: number, main: boolean): number[] {
  if (atrVal <= 0 || price <= 0) return [];
  const dir = score >= 0 ? 1 : -1;
  const strength = Math.min(1, Math.abs(score));
  const step = (atrVal / price) * (0.45 + 0.75 * strength);
  const wiggle = main ? 0.12 : 0.22;
  const out: number[] = [];
  let acc = 0;
  for (let i = 0; i < pathLen; i++) {
    const t = i / Math.max(1, pathLen - 1);
    const curve = (1 / (1 + Math.exp(-6 * (t - 0.35)))) * (1 - 0.35 * t);
    const noise = Math.sin((i + 1) * 1.7) * wiggle * step;
    acc += dir * step * curve + noise;
    out.push(acc);
  }
  return out;
}

export function analyzeTyronPro(candles: Candle[], rsiLen = 14, pathLen = 18): TyronProResult {
  if (candles.length < Math.max(60, rsiLen + 10)) {
    return {
      bias: 'NEUTRAL',
      confidence: 0,
      absorbBull: false,
      absorbBear: false,
      reasons: ['데이터 부족'],
      pathMain: [],
      pathAlt: [],
    };
  }

  const last = candles[candles.length - 1]!;
  const atrVal = atr(candles, 14);
  const body = Math.abs(last.close - last.open);
  const upperWick = last.high - Math.max(last.open, last.close);
  const lowerWick = Math.min(last.open, last.close) - last.low;

  const bullAbsorb = atrVal > 0 &&
    lowerWick >= atrVal * 0.45 &&
    lowerWick >= body * 1.2 &&
    last.close >= last.open + body * 0.35;
  const bearAbsorb = atrVal > 0 &&
    upperWick >= atrVal * 0.45 &&
    upperWick >= body * 1.2 &&
    last.close <= last.open - body * 0.35;

  const rsiCur = rsi(candles, rsiLen);
  const rsiPrev = rsi(candles.slice(0, -1), rsiLen);
  const rsiSlope = rsiCur - rsiPrev;

  const vols = candles.slice(Math.max(0, candles.length - 40)).map(c => c.volume).sort((a, b) => a - b);
  const medV = vols.length ? vols[Math.floor(vols.length / 2)]! : 0;
  const volSpike = medV > 0 ? last.volume / medV : 1;

  let score = 0;
  const reasons: string[] = [];

  if (bullAbsorb) {
    score += 0.55;
    reasons.push('아래꼬리 흡수(롱)');
  }
  if (bearAbsorb) {
    score -= 0.55;
    reasons.push('윗꼬리 흡수(숏)');
  }
  if (Math.abs(rsiSlope) > 0.8) {
    score += rsiSlope > 0 ? 0.12 : -0.12;
    reasons.push(rsiSlope > 0 ? 'RSI 모멘텀↑' : 'RSI 모멘텀↓');
  }
  if (volSpike >= 1.35) {
    score += score >= 0 ? 0.1 : -0.1;
    reasons.push('거래량 스파이크');
  }

  const bias: TyronProResult['bias'] = score >= 0.18 ? 'LONG' : score <= -0.18 ? 'SHORT' : 'NEUTRAL';
  const conf = Math.max(0, Math.min(100, Math.round((Math.min(1, Math.abs(score) / 0.85)) * 100)));

  if (reasons.length === 0) reasons.push('중립(근거 약함)');

  const pathMain = buildPath(pathLen, atrVal, last.close, score, true);
  const pathAlt = buildPath(Math.max(10, Math.round(pathLen * 0.7)), atrVal, last.close, score, false);

  return {
    bias,
    confidence: conf,
    absorbBull: bullAbsorb,
    absorbBear: bearAbsorb,
    reasons: reasons.slice(0, 4),
    pathMain,
    pathAlt,
  };
}

File: backups/server-pull-20260822-132009/extract/ailongshort/lib/logic/tyron_pro_engine.dart
import 'dart:math';
import '../data/models/candle.dart';

class TyronProResult {
  final String bias; // LONG / SHORT / NEUTRAL
  final int confidence; // 0..100
  final bool absorbBull;
  final bool absorbBear;
  final List<String> reasons; // compact bullets
  final List<double> pathMain; // relative returns (0.0..)
  final List<double> pathAlt;

  const TyronProResult({
    required this.bias,
    required this.confidence,
    required this.absorbBull,
    required this.absorbBear,
    required this.reasons,
    required this.pathMain,
    required this.pathAlt,
  });
}

/// ?€?´ë¡± PRO v2 (ê°œí¸):
/// - ê¼¬ë¦¬ ?¡ìˆ˜(engulf/absorb) + RSI ëª¨ë©˜?€ + ë³¼ë¥¨ ?¤íŒŒ?´í¬(ê°„ì´) ê¸°ë°˜
/// - ê²°ê³¼: bias/conf + "ë¯¸ë˜ ê²½ë¡œ(?ì„ )"???ë? ?˜ìµë¥??œí€€??///
/// NOTE: ë³??”ì§„?€ 'ê²°ì •??zip' ?„ë¡œ?íŠ¸?ì„œ ?ˆì „?˜ê²Œ ?™ì‘?˜ë„ë¡?///       ?¸ë? ?˜ì¡´???†ì´(?¼ì´ë¸ŒëŸ¬ë¦??†ì´) êµ¬í˜„??
class TyronProEngine {
  static TyronProResult analyze(List<Candle> candles, {int rsiLen = 14, int pathLen = 18}) {
    if (candles.length < max(60, rsiLen + 10)) {
      return const TyronProResult(
        bias: 'NEUTRAL',
        confidence: 0,
        absorbBull: false,
        absorbBear: false,
        reasons: <String>['?°ì´??ë¶€ì¡?],
        pathMain: <double>[],
        pathAlt: <double>[],
      );
    }

    final last = candles.last;
    final atr = _atr(candles, 14);
    final body = (last.c - last.o).abs();
    final upperWick = last.h - max(last.o, last.c);
    final lowerWick = min(last.o, last.c) - last.l;

    // Absorption (wick dominance + close recovery)
    final bool bullAbsorb = atr > 0 &&
        lowerWick >= atr * 0.45 &&
        lowerWick >= body * 1.2 &&
        last.c >= (last.o + body * 0.35); // closes not too low
    final bool bearAbsorb = atr > 0 &&
        upperWick >= atr * 0.45 &&
        upperWick >= body * 1.2 &&
        last.c <= (last.o - body * 0.35); // closes not too high

    // RSI momentum + micro divergence proxy
    final rsi = _rsi(candles, rsiLen);
    final rsiPrev = _rsi(candles.sublist(0, candles.length - 1), rsiLen);
    final rsiSlope = rsi - rsiPrev;

    // Volume spike proxy (relative to median of recent 30)
    final vols = candles.skip(max(0, candles.length - 40)).map((e) => e.v).toList()..sort();
    final medV = vols.isEmpty ? 0.0 : vols[vols.length ~/ 2];
    final volSpike = medV > 0 ? (last.v / medV) : 1.0;

    // Score assembly
    double score = 0.0;
    final reasons = <String>[];

    if (bullAbsorb) {
      score += 0.55;
      reasons.add('?„ë˜ê¼¬ë¦¬ ?¡ìˆ˜(ë¡?');
    }
    if (bearAbsorb) {
      score -= 0.55;
      reasons.add('?—ê¼¬ë¦??¡ìˆ˜(??');
    }

    if (rsiSlope.abs() > 0.8) {
      score += rsiSlope > 0 ? 0.12 : -0.12;
      reasons.add(rsiSlope > 0 ? 'RSI ëª¨ë©˜?€?? : 'RSI ëª¨ë©˜?€??);
    }

    if (volSpike >= 1.35) {
      score += score >= 0 ? 0.10 : -0.10;
      reasons.add('ê±°ë˜???¤íŒŒ?´í¬');
    }

    // Bias + confidence
    final bias = score >= 0.18 ? 'LONG' : (score <= -0.18 ? 'SHORT' : 'NEUTRAL');
    final conf = (min(1.0, score.abs() / 0.85) * 100.0).round().clamp(0, 100);

    if (reasons.isEmpty) reasons.add('ì¤‘ë¦½(ê·¼ê±° ?½í•¨)');

    // Build future path as relative returns using ATR and score (safe deterministic)
    final pathMain = _buildPath(pathLen: pathLen, atr: atr, price: last.c, score: score, main: true);
    final pathAlt = _buildPath(pathLen: max(10, (pathLen * 0.7).round()), atr: atr, price: last.c, score: score, main: false);

    return TyronProResult(
      bias: bias,
      confidence: conf,
      absorbBull: bullAbsorb,
      absorbBear: bearAbsorb,
      reasons: reasons.take(4).toList(),
      pathMain: pathMain,
      pathAlt: pathAlt,
    );
  }

  static List<double> _buildPath({required int pathLen, required double atr, required double price, required double score, required bool main}) {
    if (atr <= 0 || price <= 0) return const <double>[];
    final dir = score >= 0 ? 1.0 : -1.0;
    final strength = min(1.0, score.abs());
    final step = (atr / price) * (0.45 + 0.75 * strength);
    final wiggle = main ? 0.12 : 0.22;

    final out = <double>[];
    double acc = 0.0;
    for (int i = 0; i < pathLen; i++) {
      // S-curve acceleration then flatten
      final t = i / max(1, pathLen - 1);
      final curve = (1 / (1 + exp(-6 * (t - 0.35)))) * (1 - 0.35 * t);
      final noise = (sin((i + 1) * 1.7) * wiggle) * step;
      acc += dir * step * curve + noise;
      out.add(acc);
    }
    return out;
  }

  static double _atr(List<Candle> c, int len) {
    if (c.length < len + 2) return 0.0;
    final start = c.length - len;
    double sum = 0.0;
    for (int i = start; i < c.length; i++) {
      final cur = c[i];
      final prevClose = c[i - 1].c;
      final tr = max(cur.h - cur.l, max((cur.h - prevClose).abs(), (cur.l - prevClose).abs()));
      sum += tr;
    }
    return sum / len;
  }

  static double _rsi(List<Candle> c, int len) {
    if (c.length < len + 2) return 50.0;
    double gain = 0.0;
    double loss = 0.0;
    final start = c.length - len;
    for (int i = start; i < c.length; i++) {
      final diff = c[i].c - c[i - 1].c;
      if (diff >= 0) {
        gain += diff;
      } else {
        loss += -diff;
      }
    }
    if (gain == 0 && loss == 0) return 50.0;
    if (loss == 0) return 100.0;
    final rs = gain / loss;
    return 100.0 - (100.0 / (1.0 + rs));
  }
}
